simulator/simulator.go

```diff
@@ -0,0 +1,329 @@
+package simulator
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/launcher/service"
+	"github.com/kolide/osquery-go/plugin/distributed"
+	"github.com/kolide/osquery-go/plugin/logger"
+	"github.com/pkg/errors"
+)
+
+// QueryRunner is the interface which defines the pluggable behavior of a simulated
+// host. Each "type" of host may have their own implementation of this interface.
+type QueryRunner interface {
+	RunQuery(sql string) (results []map[string]string, err error)
+}
+
+// HostSimulation is the type which contains the state of a simulated host
+type HostSimulation struct {
+	// the following define the configurable aspects of the simulation
+	host                   QueryRunner
+	uuid                   string
+	enrollSecret           string
+	grpcURL                string
+	insecure               bool
+	insecureGrpc           bool
+	requestQueriesInterval time.Duration
+	requestConfigInterval  time.Duration
+	publishLogsInterval    time.Duration
+
+	// The state of the simulation is gated with a read/write lock.
+	// To read something in state:
+	//
+	//   h.state.lock.RLock()
+	//   defer h.state.lock.RUnlock()
+	//
+	// To write state based on the on-going simulation:
+	//
+	//   h.state.lock.Lock()
+	//   defer h.state.lock.Unlock()
+	state *hostSimulationState
+
+	shutdown chan bool
+	done     chan bool
+}
+
+// Enroll is the implementation of the host simulation's enrollment functionality
+func (h *HostSimulation) Enroll() error {
+	h.state.lock.Lock()
+	defer h.state.lock.Unlock()
+
+	nodeKey, invalid, err := h.state.serviceClient.RequestEnrollment(context.Background(), h.enrollSecret, h.uuid)
+	if err != nil {
+		return errors.Wrap(err, "transport error in enrollment")
+	}
+	if invalid {
+		return errors.New("enrollment invalid")
```

Comment:
this message is too generic to see where the error came from. maybe `simulator: enrollment invalid` and `fmt.Errorf` to include some information about which host failed.

Reply:
We currently don't log this error in any case... Maybe we should in `LaunchSimulation`?